File: source/cli/fdscli/model/statistics/metrics.py

class Metric(object):
    '''
    Created on Jun 29, 2015

    just a string collection of valid metric types 
    
    @author: nate
    '''
    GETS = ("GETS", "Gets")
    PUTS = ("PUTS", "Puts")
    #   gets from SSD and cache
    SSD_GETS = ( "SSD_GETS", "SSD Gets" )
    
    QFULL = ("QFULL", "Queue Full" )
    MBYTES = ( "MBYTES", "Metadata Bytes" )
    BLOBS = ( "BLOBS", "Blobs" )
    OBJECTS = ( "OBJECTS", "Objects" )
    ABS = ( "ABS", "Ave Blob Size" )
    AOPB = ( "AOPB", "Ave Objects per Blob" )
    LBYTES = ( "LBYTES", "Logical Bytes" )
    PBYTES = ( "PBYTES", "Physical Bytes" )
#    firebreak metrics
    STC_SIGMA = ( "STC_SIGMA", "Short Term Capacity Sigma" )
    LTC_SIGMA = ( "LTC_SIGMA", "Long Term Capacity Sigma" )
    STP_SIGMA = ( "STP_SIGMA", "Short Term Perf Sigma" )
    LTP_SIGMA = ( "LTP_SIGMA", "Long Term Perf Sigma" )
#   Performance IOPS
    STC_WMA = ( "STC_WMA", "Short Term Capacity WMA" )
    LTC_WMA = ( "LTC_WMA", "Long Term Capacity WMA" )
    STP_WMA = ( "STP_WMA", "Short Term Perf WMA" )
    LTP_WMA = ( "LTP_WMA", "Long Term Perf WMA" )
    
    DICT = dict()
    
    DICT[GETS[0]] = GETS
    DICT[PUTS[0]] = PUTS
    DICT[SSD_GETS[0]] = SSD_GETS
    DICT[QFULL[0]] = QFULL
    DICT[MBYTES[0]] = MBYTES
    DICT[BLOBS[0]] = BLOBS
    DICT[OBJECTS[0]] = OBJECTS
    DICT[ABS[0]] = ABS
    DICT[AOPB[0]] = AOPB
    DICT[LBYTES[0]] = LBYTES
    DICT[PBYTES[0]] = PBYTES
    DICT[STC_SIGMA[0]] = STC_SIGMA
    DICT[LTC_SIGMA[0]] = LTC_SIGMA
    DICT[STP_SIGMA[0]] = STP_SIGMA
    DICT[LTP_SIGMA[0]] = LTP_SIGMA
    DICT[STC_WMA[0]] = STC_WMA
    DICT[LTC_WMA[0]] = LTC_WMA
    DICT[STP_WMA[0]] = STP_WMA
    DICT[LTP_WMA[0]] = LTP_WMA

File: source/java/fds-project/fds/src/main/java/com/formationds/commons/model/type/Metrics.java
/*
 * Copyright (c) 2014, Formation Data Systems, Inc. All Rights Reserved.
 */

package com.formationds.commons.model.type;

import com.formationds.commons.model.exception.UnsupportedMetricException;
import com.formationds.commons.model.i18n.ModelResource;

import java.util.EnumSet;

/**
 * @author ptinius
 */
public enum Metrics {
    PUTS( "Puts" ),
    GETS( "Gets" ),
    QFULL( "Queue Full" ),
    /**
     * @deprecated use {@link #SSD_GETS}
     */
    @Deprecated
    PSSDA( "Percent of SSD Accesses" ),
    MBYTES( "Metadata Bytes" ),
    BLOBS( "Blobs" ),
    OBJECTS( "Objects" ),
    ABS( "Ave Blob Size" ),
    AOPB( "Ave Objects per Blob" ),
    UBYTES( "Used Bytes" ),
    LBYTES( "Logical Bytes" ),
    PBYTES( "Physical Bytes" ),
    // firebreak metrics
    STC_SIGMA( "Short Term Capacity Sigma" ),
    LTC_SIGMA( "Long Term Capacity Sigma" ),
    STP_SIGMA( "Short Term Perf Sigma" ),
    LTP_SIGMA( "Long Term Perf Sigma" ),
    // Performance IOPS
    STC_WMA( "Short Term Capacity WMA" ),
    LTC_WMA( "Long Term Capacity WMA" ),
    STP_WMA( "Short Term Perf WMA" ),
    LTP_WMA( "Long Term Perf WMA" ),
    /**
     * gets from SSD and cache
     */
    SSD_GETS( "SSD Gets" ),
    /**
     * Gets from just HDD
     */
    HDD_GETS( "HDD Gets" );

    public static EnumSet<Metrics> FIREBREAK = EnumSet.of(Metrics.STC_SIGMA,
                                                          Metrics.LTC_SIGMA,
                                                          Metrics.STP_SIGMA,
                                                          Metrics.LTP_SIGMA);

    public static EnumSet<Metrics> FIREBREAK_CAPACITY = EnumSet.of(Metrics.STC_SIGMA,
                                                                   Metrics.LTC_SIGMA);

    public static EnumSet<Metrics> FIREBREAK_PERFORMANCE = EnumSet.of(Metrics.STP_SIGMA,
                                                                      Metrics.LTP_SIGMA);

    public static EnumSet<Metrics> PERFORMANCE = EnumSet.of(Metrics.STP_WMA);

    public static final EnumSet<Metrics> CAPACITY = EnumSet.of(Metrics.PBYTES,
                                                               Metrics.LBYTES);
    
    public static final EnumSet<Metrics> PERFORMANCE_BREAKDOWN = EnumSet.of( Metrics.PUTS,
    																		 Metrics.HDD_GETS,
    																		 Metrics.SSD_GETS);

    private static final String UNKNOWN_METRIC =
        ModelResource.getString( "model.metrics.unsupported" );

    private final String key;

    /**
     * @param key the {@link String} representing the metadata key
     */
    Metrics( String key ) {
        this.key = key;
    }

    /**
     * @return Returns{@link String} representing the metadata key
     */
    public String key() {
        return this.key;
    }
    
    /**
     * Takes a string and checks it against both the name and the metadata key
     *
     * @param aValue the value to check
     *
     * @return true if the value matches either the name or the key, ignoring case
     */
    public boolean matches( String aValue ) {
    	
    	return ( aValue.equalsIgnoreCase( this.key() ) || aValue.equalsIgnoreCase( this.name() ) );
    }

    /**
     * Lookup the metric by the name or the metadata key name.  This does a case-insensitive comparison
     * against the metadata key and the metric enum name
     *
     * @param metadataKey the case sensitive {@link String} representing the
     *                    metric name OR the metadata key name
     *
     * @return Returns {@link com.formationds.commons.model.type.Metrics}
     * associated with {@code metadataKey}
     *
     * @throws UnsupportedMetricException if the the {@code metadataKey} is not found
     */
    public static Metrics lookup( final String metadataKey )
        throws UnsupportedMetricException {
        String tmdk = metadataKey.trim();
        for( final Metrics m : values() ) {
            if( m.matches( tmdk ) ) {
                return m;
            }
        }

        throw new UnsupportedMetricException( String.format( UNKNOWN_METRIC,
                                                             metadataKey ) );
    }

}
